ts/src/profiler.ts

```diff
@@ -201,29 +254,37 @@ export class Profiler extends common.ServiceObject {
     const delayMillis = await this.collectProfile();
 
     // Schedule the next profile.
-    setTimeout(this.runLoop.bind(this), delayMillis).unref();
+    if (typeof delayMillis === 'number') {
+      setTimeout(this.runLoop.bind(this), delayMillis).unref();
+      return;
+    }
+
+    setTimeout(this.runLoop.bind(this), delayMillis.getBackoff()).unref();
   }
 
   /**
    * Waits for profiler server to tell it to collect a profile, then collects
    * a profile and uploads it.
    *
    * @return - time, in ms, to wait before asking profiler server again about
-   * collecting another profile.
-   *
-   * TODO: implement backoff and retry. When error encountered in
-   * createProfile() should be retried when response indicates this request
-   * should be retried or with exponential backoff (up to one hour) if the
-   * response does not indicate when to retry this request.
+   * collecting another profile, or retryer specifying exponential backoff.
    */
-  async collectProfile(): Promise<number> {
+  async collectProfile(retryer?: Retryer): Promise<number|Retryer> {
```

Comment:
This chaining / piping of retryers looks too complicated. Why not just have the retrying object as a member of the class, have a reset() function and reset the backoff on successful profile creation?

Reply:
Done.